File: medium_load.cs
using System;
using System.Diagnostics;

class MediumLoad
{
    static void Swap(int[] arr, int i, int j)
    {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    static int Partition(int[] arr, int low, int high)
    {
        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++)
        {
            if (arr[j] <= pivot)
            {
                i++;
                Swap(arr, i, j);
            }
        }

        Swap(arr, i + 1, high);
        return i + 1;
    }

    static void QuickSort(int[] arr, int low, int high)
    {
        if (low < high)
        {
            int pi = Partition(arr, low, high);
            QuickSort(arr, low, pi - 1);
            QuickSort(arr, pi + 1, high);
        }
    }

    static void PrintArray(int[] arr)
    {
        Console.WriteLine($"First element: {arr[0]}");
        Console.WriteLine($"Last element: {arr[arr.Length - 1]}");
    }

    static void Main(string[] args)
    {
        long n = 1_000_000;
        int[] nums = new int[n];
        int min = 1, max = 100;

        Random rand = new Random();
        Console.WriteLine("Sorting 10000 random numbers from 1 to 100...");

        for (int i = 0; i < n; i++)
            nums[i] = rand.Next(min, max + 1);

        PrintArray(nums);

        var startTimestamp = Stopwatch.GetTimestamp();
        Console.WriteLine($"Start time: {(double)startTimestamp / Stopwatch.Frequency}");

        var sw = Stopwatch.StartNew();

        QuickSort(nums, 0, nums.Length - 1);

        sw.Stop();
        var endTimestamp = Stopwatch.GetTimestamp();

        PrintArray(nums);

        Console.WriteLine($"End time: {(double)endTimestamp / Stopwatch.Frequency}");
        Console.WriteLine($"Total time: {sw.Elapsed.TotalSeconds:F3}");

        Console.Out.Flush();
    }
}

File: medium_load.js
// Node.js program: quicksort random integers

function swap(arr, i, j) {
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

function partition(arr, low, high) {
    const pivot = arr[high];
    let i = low - 1;
    for (let j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr, i, j);
        }
    }
    swap(arr, i + 1, high);
    return i + 1;
}

function quickSort(arr, low, high) {
    if (low < high) {
        const pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

function printArray(arr) {
    console.log(`First element: ${arr[0]}`);
    console.log(`Last element: ${arr[arr.length - 1]}`);
}

// ---- main ----
const n = 1_000_000;
const nums = new Array(n);
const min = 1;
const max = 100;

console.log("Sorting 1,000,000 random numbers from 1 to 100...");

// generate random numbers
for (let i = 0; i < n; i++) {
    nums[i] = Math.floor(Math.random() * (max - min + 1)) + min;
}

let start = process.hrtime.bigint();
console.log(`Start time: ${(Number(start) / 1e9).toFixed(3)}`);

printArray(nums);
quickSort(nums, 0, n - 1);

let end = process.hrtime.bigint();

printArray(nums);

console.log(`End time: ${(Number(end) / 1e9).toFixed(3)}`);
console.log(`Total time: ${(Number(end - start) / 1e9).toFixed(3)}`);
